n3fit/src/n3fit/io/writer.py

```diff
@@ -213,98 +218,156 @@
 
-    def write_data(self, replica_path_set, fitname, tr_chi2, vl_chi2, true_chi2):
+    def write_data(self, save_path, fitname, weights_name):
         """
-        Wrapper around the `storefit` function.
+        Save all the data of a fit, for all replicas.
 
         Parameters
         ----------
-            `replica_path_set`
-                full path for the replica, ex: `${PWD}/runcard_name/nnfit/replica_1`
+            `save_path`
+                path for the replica, ex: `${PWD}/runcard_name/nnfit`
             `fitname`
-                name of the fit
-            `tr_chi2`
-                training chi2
-            `vl_chi2`
-                validation chi2
-            `true_chi2`
-                chi2 of the replica to the central experimental data
+                name of the fit, ex: `Basic_runcard`
+            `weights_name`
+                name of the file to save weights to, if not empty
         """
-        # Check the directory exist, if it doesn't, generate it
-        os.makedirs(replica_path_set, exist_ok=True)
+        os.makedirs(save_path, exist_ok=True)
+
+        self.preprocessing = [
+            pdf_object.get_preprocessing_factors() for pdf_object in self.pdf_objects
+        ]
+        self.arc_lengths = [
+            vpinterface.compute_arclength(pdf_object).tolist() for pdf_object in self.pdf_objects
+        ]
+        self.integrability_numbers = [
+            vpinterface.integrability_numbers(pdf_object).tolist()
+            for pdf_object in self.pdf_objects
+        ]
+
+        for i in range(len(self.replica_numbers)):
+            replica_path = f"{save_path}/replica_{self.replica_numbers[i]}"
+            os.makedirs(replica_path, exist_ok=True)
+
+            self._write_chi2s(f"{replica_path}/chi2exps.log")
+            self._write_metadata_json(i, f"{replica_path}/{fitname}.json")
+            self._export_pdf_grid(i, f"{replica_path}/{fitname}.exportgrid")
+            if weights_name:
+                self._write_weights(i, f"{replica_path}/{weights_name}")
+
+    def _write_chi2s(self, out_path):
+        # Note: same for all replicas, unless run separately
+        chi2_log = self.stopping_object.chi2exps_json()
+        with open(out_path, "w", encoding="utf-8") as fs:
+            json.dump(chi2_log, fs, indent=2, cls=SuperEncoder)
 
-        stop_epoch = self.stopping_object.stop_epoch
+    def _write_metadata_json(self, i, out_path):
+        json_dict = jsonfit(
+            best_epoch=self.stopping_object.best_epochs[i],
+            positivity_status=self.stopping_object.positivity_statusses[i],
+            preprocessing=self.preprocessing[i],
+            arc_lengths=self.arc_lengths[i],
+            integrability_numbers=self.integrability_numbers[i],
+            tr_chi2=self.tr_chi2[i],
+            vl_chi2=self.vl_chi2[i],
+            true_chi2=self.true_chi2[i],
+            # Note: last 2 same for all replicas, unless run separately
```

Comment:
What does this comment exactly mean? These two does not necessarily have to be the same even if the replicas were fitted in parallel (if that's what you were referring to).

Reply:
The "last 2" refers to the two lines below the comment, `timing` and `stop_epoch`. If running multiple replicas in parallel, these are the overall values over all replicas. It's not an issue, just thought it'd be useful to document here.
Shall I just add the word "below" to make it clear?

Reply:
Yes, adding "below" will definitely make it clearer.